yt/yt/library/decimal/decimal.cpp

```diff
@@ -26,10 +144,138 @@ struct TDecimalTraits
+        return static_cast<TU>(value);
+    }
+}
+
+template <typename T>
+Y_FORCE_INLINE constexpr auto DecimalIntegerToSigned(T value)
+{
+    static_assert(ValidDecimalUnderlyingUnsignedInteger<T>);
+
+    if constexpr (std::is_same_v<T, ui256>) {
+        return i256{value.Parts};
+    } else if constexpr (std::is_same_v<T, ui128>) {
+        return i128(value);
+    } else {
+        using TU = std::make_signed_t<T>;
+        return static_cast<TU>(value);
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+//! The functions below are used for implementing conversion to/from text to binary decimal.
+//! We actually do not need any arithmetic operations beside v *= 10, v /= 10, addition and negation.
+//! so they are the only ones actually implemented for our custom i256/ui256.
+
+template <typename T>
+Y_FORCE_INLINE constexpr auto MSBFlip(T value)
+{
+    static_assert(ValidDecimalUnderlyingInteger<T>);
+
+    if constexpr (std::is_same_v<T, i256>) {
+        value.Parts[7] ^= (1u << 31);
+        return value;
+    } else {
+        constexpr auto one = DecimalIntegerToUnsigned(T{1});
+        // Bit operations are only valid with unsigned types.
+        return T(DecimalIntegerToUnsigned(value) ^ (one << (sizeof(T) * 8 - 1)));
+    }
+}
+
+template <typename T>
+Y_FORCE_INLINE constexpr ui32 GetNextDigit(T value, T* nextValue)
+{
+    static_assert(ValidDecimalUnderlyingUnsignedInteger<T>);
+
+    if constexpr (std::is_same_v<T, ui256>) {
+        ui64 remainder = 0;
+        for (int partIndex = 7; partIndex >= 0; --partIndex) {
+            // Everything should fit into long long since we are dividing by 10.
+            auto step = std::lldiv(value.Parts[partIndex] + (remainder << 32), 10);
+            value.Parts[partIndex] = step.quot;
+            remainder = step.rem;
+        }
+        *nextValue = value;
+        return remainder;
+    } else {
+        constexpr auto ten = T{10};
+        *nextValue = value / ten;
+        return static_cast<ui32>(value % ten);
+    }
+}
+
+template <typename T>
+Y_FORCE_INLINE constexpr T MultiplyByTen(T value)
+{
+    static_assert(ValidDecimalUnderlyingUnsignedInteger<T>);
+
+    if constexpr (std::is_same_v<T, ui256>) {
+        // 2 * (4 * v + v) = 10v.
+        return ShiftUp<1>(ShiftUp<2>(value) + value);
+    } else {
+        return value * DecimalIntegerToUnsigned(10);
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+TDecimal::TValue256 Value128ToValue256(TDecimal::TValue128 value)
+{
+    TDecimal::TValue256 result;
+    std::memcpy(&result, &value, sizeof(value));
+    std::fill(
+        result.Parts.begin() + 4,
+        result.Parts.end(),
+        value.High < 0 ? std::numeric_limits<ui32>::max() : 0u);
+    return result;
+}
+
+TDecimal::TValue128 Value256ToValue128(const TDecimal::TValue256& value)
+{
+    TDecimal::TValue128 result;
+    std::memcpy(&result, &value, sizeof(result));
```

Comment:
Are you intentionally copy highest but now lowest bits here?

Reply:
Yes, surprisingly, it is the correct way. But I guess I will remove this altogether after ermolovd's comments.